### PR TITLE
core/state: maintain one more diff layer

### PR DESCRIPTION
**Pull request by rjl493456442**
It's a PR for #21724

### CODE REVIEW

File: core/state/statedb.go [hunk 1]
```diff
@@ -869,7 +869,11 @@ func (s *StateDB) Commit(deleteEmptyObjects bool) (common.Hash, error) {
 			if err := s.snaps.Update(root, parent, s.snapDestructs, s.snapAccounts, s.snapStorage); err != nil {
 				log.Warn("Failed to update snapshot tree", "from", parent, "to", root, "err", err)
 			}
-			if err := s.snaps.Cap(root, 127); err != nil { // Persistent layer is 128th, the last available trie
+			// Keep 128 diff layers in the memory, persistent layer is 129th.
+			// - head layer is paired with HEAD state
+			// - head-1 layer is paired with HEAD-1 state
+			// - head-127 layer(bottom-most diff layer) is paired with HEAD-127 state
+			if err := s.snaps.Cap(root, 128); err != nil {
 				log.Warn("Failed to cap snapshot tree", "root", root, "layers", 127, "err", err)
```
Comment (karalabe) [inline core/state/statedb.go:877]: 127->128 in the logs too pls

**Comment by karalabe** [comment]
What's the reason for needing +1 state?

**Comment by rjl493456442** [comment]
@karalabe 

Because in the pruner, we decide to use `HEAD-127` as the pruning target.
The reason for picking this number is: this state is available(in most of the
normal cases) and this state is old enough to prevent some tiny reorg.

However now we only maintain 128 layers in total. 127 diff layers + 1 disk layer.
It means head-126 is the bottom-most diff layer. While head-127 diff layer is
already merged into the `head-126`. So `HEAD-127` state has no associated
snapshot.

**Comment by karalabe** [comment]
Good point

File: core/state/statedb.go (hunk 1 above)
Comment (karalabe) [reply core/state/statedb.go:877]: Ping?

Comment (rjl493456442) [reply core/state/statedb.go:877]: Sorry, Fixed!

**Review by karalabe — APPROVED** [review]
LGTM

**Pull request merged**